packages/snap/src/index.ts

```diff
@@ -0,0 +1,463 @@
+import { Address } from "./types/address";
+import { ChainState, AddressState } from "./state";
+import { Result } from "./types/result";
+import { submitTransaction } from "./transaction";
+import { getAddress } from "./address";
+
+/**
+ * Handle incoming JSON-RPC requests, sent through `wallet_invokeSnap`.
+ *
+ * @param args - The request handler args as object.
+ * @param args.request - A validated JSON-RPC request object.
+ * @returns A result object.
+ * @throws If the request method is not valid for this snap.
+ */
+export const onRpcRequest: OnRpcRequestHandler = async ({
+  request,
+}): Promise<Result> => {
+  let res: Object = {};
+  let confirmation: string | boolean | null = false;
+  switch (request.method) {
+    case "initialize":
+      // Ensure user confirms initializing Cosmos snap
+      confirmation = await snap.request({
+        method: "snap_dialog",
+        params: {
+          type: "confirmation",
+          content: panel([
+            text(
+              "Would you like to add Cosmos chain support within your Metamask wallet?"
+            ),
+          ]),
+        },
+      });
+      if (!confirmation) {
+        throw new Error("Initialize Cosmos chain support was denied.");
+      }
+      let chains = new Chains([]);
+      let chainList = await initializeChains();
+      chains = new Chains(chainList);
+      // Initialize with initial state
+      await snap.request({
+        method: "snap_manageState",
+        params: {
+          operation: "update",
+          newState: { chains: chains.string(), addresses: JSON.stringify([]) },
+        },
+      });
+
+      await snap.request({
+        method: "snap_dialog",
+        params: {
+          type: "alert",
+          content: panel([
+            heading("Initialization Successful"),
+            text(
+              "Cosmos has been added and initialized into your Metamask wallet."
+            ),
+          ]),
+        },
+      });
+
+      return {
+        data: res,
+        success: true,
+        statusCode: 201,
+      };
+    case "transact":
+      // Send a transaction to the wallet
+      if (
+        !(
+          request.params != null &&
+          typeof request.params == "object" &&
+          "msgs" in request.params &&
+          "chain_id" in request.params &&
+          typeof request.params.msgs == "string" &&
+          typeof request.params.chain_id == "string"
+        )
+      ) {
+        throw new Error("Invalid transact request");
+      }
+
+      // Ensure user confirms transaction
+      confirmation = await snap.request({
+        method: "snap_dialog",
+        params: {
+          type: "confirmation",
+          content: panel([
+            heading("Confirm Transaction"),
+            divider(),
+            heading("Chain"),
+            text(`${request.params.chain_id}`),
+            divider(),
+            heading("Transaction"),
+            text(`${request.params.msgs}`),
```

Comment:
Since you're submitting the JSON-parsed messages below, I suggest rendering what will be submitted:
```js
const txMessages = JSON.parse(request.params.msgs);
```
```suggestion
            text(JSON.stringify(txMessages, null, 2)),
```